crates/sui-indexer/src/models/checkpoints.rs

```diff
@@ -118,6 +116,13 @@ impl TryFrom<StoredCheckpoint> for RpcCheckpoint {
                 })
                 .collect::<Result<Vec<TransactionDigest>, IndexerError>>()?
         };
+
+        let max_tx_sequence_number = checkpoint.max_tx_sequence_number.ok_or_else(|| {
+            IndexerError::PersistentStorageDataCorruptionError(
+                "max_tx_sequence_number should not be None".to_string(),
```

Comment:
@amnn good callout regarding empty checkpoint!

> I notice a lot of adding +1/-1 

indeed this is mainly b/c both upstream checkpoint data and JSON RPC checkpoint need `network_total_transactions` while we only stored `max_tx_sequence_number`, which less than `network_total_transactions` by 1.

```
tx_lo = 1 + max(tx for checkpoint c < C for tx in c)
tx_hi = 1 + max(tx for checkpoint c <= C for tx in c) 
```

this would work indeed, while I feel it's a bit hard to grasp the meaning of tx_lo & tx_hi intuitively, another alternative that would also cover the empty checkpoint case would be

- network_total_transactions for tx count at the end of the checkpoint, this is also in checkpoint;
- checkpoint_total_transactions for tx count of current checkpoint, which is 0 for empty case

and then we can derive min & max, now mainly for pruning and change the meaning like to half-open orthogonally if we want, wdyt?

Reply:
The counts idea is interesting -- I quite like it -- the only thing to check is whether we ever need to filter on `min_tx_sequence_number`, because if so, you want to store that directly, rather than compute it (so that you can build an index over it).

FWIW, I think what makes the `tx_lo`/`tx_hi` description non-intuitive is trying to make it correct in all cases (particularly the empty case), but there is a fairly natural intuitive description of it, which is that what we're doing is that we are tracking the boundaries between checkpoints always (with the upperbound of one checkpoint equalling the lowerbound of the next).

Thinking in those terms, using half open intervals is the easiest way to capture all the edge cases, because it can represent an empty checkpoint while still remembering where in the stream of checkpoints/transactions it belongs.

Reply:
> whether we ever need to filter on min_tx_sequence_number

I like the forward thinking! although I think for the tx seq -> cp seq lookup, we will want to leverage transaction tables instead of building an index on checkpoint table. (also the case when people have a tx seq, knowing that it's a min_tx_sequence_number and want to find the related cp seems pretty rare to me, this may be debatable), does that make sense?

Reply:
> also the case when people have a tx seq, knowing that it's a min_tx_sequence_number and want to find the related cp seems pretty rare to me, this may be debatable

The kind of query that I'm talking about is not one where the tx you hold in your hand happens to be a min tx sequence number, but one where you wanted to select checkpoints but by applying a bound on their min_tx_sequence_number.

This is not that uncommon -- we do this on the `epochs` table, bounding by `first_checkpoint_id` (for consistent read reasons). I think the broader point is that there is value in avoiding computing the actual bounds, because they are more likely to show up in filters, as opposed to the actual count of transactions, which is unlikely to show up in filters.

Reply:
the broader point makes sense, will do it that way.